File: src/clj/big_config/step_fns.clj
(ns big-config.step-fns
  (:require
   [big-config :as bc]
   [big-config.core :refer [->step-fn]]
   [bling.core :refer [bling]]
   [clojure.java.io :as io]
   [clojure.string :as str]
   [selmer.parser :as p]
   [selmer.util :as util]))

(defn exit-with-code [n]
  (shutdown-agents)
  (flush)
  (System/exit n))

(defn ->exit-step-fn [end]
  (->step-fn {:after-f (fn [step {:keys [::bc/env ::bc/exit]}]
                         (when (and (= step end)
                                    ((complement #{:repl}) env))
                           (exit-with-code exit)))}))

(defn ->print-error-step-fn [end]
  (->step-fn {:before-f (fn [step {:keys [::bc/err
                                          ::bc/exit
                                          ::bc/stack-trace]}]
                          (let [color :red.bold
                                prefix "\uf05c"
                                msg (when (and (= step end)
                                               (> exit 0)
                                               (string? err)
                                               (not (str/blank? err))) err)
                                stack-trace (when (and (= step end)
                                                       (> exit 0)
                                                       (not (str/blank? stack-trace))) stack-trace)]
                            (when msg
                              (binding [*out* *err*]
                                (println (bling [color (p/render (str "{{ prefix }} " msg) {:prefix prefix})]))))
                            (when stack-trace
                              (let [temp-file (java.io.File/createTempFile "big-config-" ".txt" (java.io.File. "/tmp"))]
                                (with-open [writer (io/writer temp-file)]
                                  (.write writer (str stack-trace)))
                                (binding [*out* *err*]
                                  (println (bling [color (p/render "\nThe stack-trace has been written to {{temp-file}}" {:temp-file (.getAbsolutePath temp-file)})])))))))}))

(def tap-step-fn
  (let [f (fn [label step opts]
            (tap> [step label opts]))]
    (->step-fn {:before-f (partial f :before)
                :after-f (partial f :after)})))

(defn log-step-fn [f step opts]
  (->> (update opts ::bc/steps (fnil conj []) step)
       (f step)))

(def bling-step-fn
  (->step-fn {:before-f (fn [step _]
                          (let [prefix "\ueabc"
                                color :green.bold]
                            (binding [*out* *err*
                                      util/*escape-variables* false]
                              (println (bling [color (p/render "{{ prefix }} {{ msg }}" {:prefix prefix
                                                                                         :msg (name step)})])))))
              :after-f (fn [step {:keys [::bc/exit
                                         ::bc/err]}]
                         (let [prefix "\uf05c"
                               color :red.bold]
                           (when (> exit 0)
                             (binding [*out* *err*
                                       util/*escape-variables* false]
                               (println (bling [color (p/render "{{ prefix }} {{ msg }}: {{ err }}" {:prefix prefix
                                                                                                     :msg (name step)
                                                                                                     :err err})]))))))}))

(comment
  [tap-step-fn bling-step-fn]
  (log-step-fn nil nil nil))

File: src/step-fns.ts
import { mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { createStepFn, type StepFn, type WrappedStep } from "./core.js";
import { ENV, ERR, EXIT, STACK_TRACE, STEPS_TRACE, type Keyword, type Opts } from "./keys.js";
import { registerFunction } from "./utils.js";

export function exitWithCode(n: number): never {
  process.exit(n);
}

export function createExitStepFn(end: Keyword): StepFn {
  return createStepFn({
    afterF: (step, opts) => {
      if (step === end && opts[ENV] !== "repl") exitWithCode(opts[EXIT] ?? 0);
    }
  });
}

export function createPrintErrorStepFn(end: Keyword): StepFn {
  return createStepFn({
    beforeF: (step, opts) => {
      const err = opts[ERR];
      const exit = opts[EXIT];
      if (step !== end || !(exit > 0)) return;
      if (typeof err === "string" && err.trim().length > 0) console.error(`✖ ${err}`);
      const stack = opts[STACK_TRACE];
      if (typeof stack === "string" && stack.trim().length > 0) {
        const dir = mkdtempSync(join(tmpdir(), "big-config-"));
        const file = join(dir, "stack-trace.txt");
        writeFileSync(file, stack);
        console.error(`\nThe stack-trace has been written to ${file}`);
      }
    }
  });
}

export const tapStepFn: StepFn = createStepFn({
  beforeF: (step, opts) => {
    const g = globalThis as any;
    if (!Array.isArray(g.__bigConfigTaps)) g.__bigConfigTaps = [];
    g.__bigConfigTaps.push([step, "before", opts]);
  },
  afterF: (step, opts) => {
    const g = globalThis as any;
    if (!Array.isArray(g.__bigConfigTaps)) g.__bigConfigTaps = [];
    g.__bigConfigTaps.push([step, "after", opts]);
  }
});

export function logStepFn(f: WrappedStep, step: Keyword, opts: Opts): Opts {
  return f(step, { ...opts, [STEPS_TRACE]: [...(opts[STEPS_TRACE] ?? []), step] });
}

export const blingStepFn: StepFn = createStepFn({
  beforeF: (step) => console.error(`➜ ${step}`),
  afterF: (step, opts) => {
    if (opts[EXIT] > 0) console.error(`✖ ${step}: ${opts[ERR] ?? ""}`);
  }
});

registerFunction("big-config.step-fns/tap-step-fn", tapStepFn as any);
registerFunction("big-config.step-fns/log-step-fn", logStepFn as any);
registerFunction("big-config.step-fns/bling-step-fn", blingStepFn as any);

export { exitWithCode as "exit-with-code", createExitStepFn as "->exit-step-fn", createPrintErrorStepFn as "->print-error-step-fn", tapStepFn as "tap-step-fn", logStepFn as "log-step-fn", blingStepFn as "bling-step-fn" };
